File: java/102.binary-tree-level-order-traversal.java
/*
 * @lc app=leetcode id=102 lang=java
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if(root == null) return new ArrayList();
        
        ArrayDeque<TreeNode> q = new ArrayDeque();
        List<List<Integer>> res = new ArrayList();
        List<Integer> currLevel = new ArrayList();

        q.push(root);
        int levelCount = 1;
        int nextLevelCount = 0;

        while(!q.isEmpty()) {
            if(levelCount == 0) {
                res.add(currLevel);
                currLevel = new ArrayList();
                levelCount = nextLevelCount;
                nextLevelCount = 0;
            }

            levelCount--;
            root = q.pollLast();

            currLevel.add(root.val);

            if(root.left != null) {
                q.push(root.left);
                nextLevelCount++;
            }
            if(root.right != null) {
                q.push(root.right);
                nextLevelCount++;
            }
        }

        res.add(currLevel);

        return res;
    }
}
// @lc code=end

File: cpp/102.binary-tree-level-order-traversal.cpp
/*
 * @lc app=leetcode id=102 lang=cpp
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        std::vector<std::vector<int>> res;
        if(!root) return res;
        std::queue<TreeNode*> q;

        q.push(root);

        int levelCount = 0;
        int nextLevelCount = 1;
        while(!q.empty()) {
            if(levelCount == 0) {
                levelCount = nextLevelCount;
                nextLevelCount = 0;
                res.emplace_back();
            }

            root = q.front();
            q.pop();
            levelCount--;
            res[res.size() - 1].push_back(root->val);

            if(root->left) {
                q.push(root->left);
                nextLevelCount++;
            }
            if(root->right) {
                q.push(root->right);
                nextLevelCount++;
            }
        }

        return res;
    }
};
// @lc code=end

